chocolatey/choco
# (#3008) Remove support for passing in nupkg file when installing or upgrading packages

## Description Of Changes

This pull request removes the support that was previously available to install a package by passing in its nupkg or nuspec file directly on the Commandline when calling `choco install` or `choco upgrade`.
Instead it will now fail the installation before it even tries to read the file by throwing an exception about it not being supported, as well as giving an example to the user how the package can be installed.
This example is only available when we reliably can determine that it is a absolute local file path, a UNC path or a relative file path where the file exists.
Otherwise a generic error output about passing in a nupkg file is shown.
If a nuspec file is shown, a different error is displayed mentioning that the file need to be packaged before attempting the installation.

## Motivation and Context

There has always been problems when installing a package when a path to the nupkg or nuspec file has been passed in, and with the uplift even more problems was discovered and it was decided to not support this feature anymore.

## Testing

Most tests can be ran without the package being available.

1. Run `choco install c:\testing\package.nupkg` (use any package, but without a version)
2. Verify the installation exits with code 1 and reports passing in a nupkg is not supported.
3. Verify the example given is along the lines of `choco install package --source="C:\testing"`
4. Run `choco install c:\testing\package.5.0.0.nupkg`
5. Verify same as number 2 and the example given is along the lines of `choco install package --version="5.0.0" --source="c:\testing"`
6. Run `choco install \\localhost\\c$\testing\package.5.0.0.nupkg`
6. Verify same as number 2, that it mentions UNC path and the example given is along the lines of `choco install package --version="5.0.0" --source="\\localhost\\c$\testing"`
7. Run `choco install c:\testing\another-package.2.0.0-alpha.nupkg`
8. Verify same as number 2, and the example given is along the lines of `choco install another-package --version="2.0.0-alpha" --prerelease --source="c:\testing"
9. Navigate into `c:\testing` and run `choco install existing-package.nupkg` (make sure this package exists)
10. Verify same as number 2, and the example given is along the lines of `choco install existing-package --source="c:\testing"`
11. Navigate into `c:\testing` and run `cohco install not-existing.nupkg` (make sure this package **DOES NOT** exist)
12. Verify a generic message about `nupkg` files passed in is not supported, and no example is given
13. Run `choco install https://testing.com/repository/some-package.nupkg`
14. Verify the same as number 12
15. Run `choco install C:\testing\meta-package.nuspec`
16. Verify an error message about the need to run `choco pack` is needed before installing the package. (No example should be given)

### Operating Systems Testing

- Windows 10/11

## Change Types Made



## Change Checklist

* [ ] Requires a change to the documentation. (Have not checked if there is any documentation changes needed)

## Related Issue

- Fixes #3008
- https://app.clickup.com/t/20540031/PROJ-464
